File: async/ts/async.ts
import fetchURL from 'node-fetch';
import { from, merge, Observable } from 'rxjs';
import { reduce } from 'rxjs/operators';

async function fetchEndpoint(url: string): Promise<number> {
    // start the fetch (header and first chunk only)
    const resp = await fetchURL(url);
    // ensure full text (may be multiple chunks)
    const text = await resp.text();
    return text.length;
}

async function runSync(url: string, count: number) {
    // invokes async function synchronously {count} times in sequence
    let bytes: number = 0;
    for (let i = 0; i < count; ++i) {
        bytes += await fetchEndpoint(url);
    }
    return bytes;
}

async function runAsync(url: string, count: number): Promise<number> {
    // starts {count} async operations concurrently, collecting their promises
    const promises = [];
    for (let i = 0; i < count; ++i) {
        promises.push(fetchEndpoint(url));
    }
    // wait for _all_ promises to complete
    const results = await Promise.all(promises);
    // sum and return
    return results.reduce((acc, n) => (acc + n), 0);
}

function runObservable(url: string, count: number): Promise<number> {
    // construct {count} observables from the {count} promises
    // (similar to runAsync)
    const requests: Array<Observable<number>> = Array(count).fill(url)
        .map((u) => from(fetchEndpoint(u)));
    // merge Array<Observable<number>> => Observable<Array<number>>
    return merge(...requests)
        // ... => Observable<number> containing sum total bytes
        .pipe(reduce((acc, x) => acc + x))
        // to Promise for async function
        .toPromise();
}

async function main() {
    const path = require('path');
    const process = require('process');
    const RequestCount: number = 100;

    if (process.argv.length < 3) {
        const myname = path.basename(process.argv[1]);
        console.log(`Usage: ${myname} URL`);
        process.exit(1);
    }
    const TargetUrl: string = process.argv[2];

    // map operation to label for data-driven operation
    const operationMap: Map<CallableFunction, string> = new Map([
        [runSync, 'Sync'],
        [runAsync, 'Async'],
        [runObservable, 'Observable'],
    ]);

    operationMap.forEach((label, f) => {
        console.log(`${label} starting...`);

        const startTime: number = Date.now();
        f(TargetUrl, RequestCount).then((bytes) => {
            const delta = Date.now() - startTime;
            console.log(`${label}: ${bytes} bytes in ${delta}ms.`);
        });
    });
}

main();

File: async/go/async.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"path"
	"sync"
	"time"
)

func fetchURL(url string) int {
	resp, err := http.Get(url)
	if err != nil {
		fmt.Println(err)
		return 0
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println(err)
		return 0
	}

	resp.Body.Close()
	return len(body)
}

func fetchSync(url string, count int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println("Sync starting...")

	start := time.Now()

	var bytes int
	for i := 0; i < count; i++ {
		bytes += fetchURL(url)
	}
	delta := time.Now().Sub(start)

	fmt.Printf("Sync: %d bytes in %s\n", bytes, delta)
}

func fetchWrapper(url string) chan int {
	c := make(chan int)

	go func() {
		defer close(c)
		c <- fetchURL(url)
	}()

	return c
}

func fetchAsync(url string, count int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println("Async starting...")

	start := time.Now()
	var chans []chan int
	for i := 0; i < count; i++ {
		chans = append(chans, fetchWrapper(url))
	}
	var bytes int
	for _, c := range chans {
		bytes += <-c
	}
	delta := time.Now().Sub(start)

	fmt.Printf("Async: %d bytes in %s\n", bytes, delta)
}

func main() {
	if len(os.Args) < 2 {
		myname := path.Base(os.Args[0])
		fmt.Printf("Usage: %s URL\n", myname)
		os.Exit(1)
	}

	targetURL := os.Args[1]

	requestCount := 500

	var wg sync.WaitGroup
	wg.Add(2)

	// kick off routines in background
	go fetchSync(targetURL, requestCount, &wg)
	go fetchAsync(targetURL, requestCount, &wg)

	wg.Wait()
}
